x-pack/plugins/fleet/server/services/agent_policies/full_agent_policy.ts

```diff
@@ -80,6 +82,23 @@ export async function getFullAgentPolicy(
 
   const sourceUri = await getSourceUriForAgentPolicy(soClient, agentPolicy);
 
+  // Build up an in-memory object for looking up Package Info, so we don't have
+  // call `getPackageInfo` for every single policy, which incurs performance costs
+  const packageInfoCache = new Map<string, PackageInfo>();
```

Comment:
We call `getPackageInfo` a LOT across the Fleet codebase, and it seems like this might be okay on the surface because we're expecting mainly cache hits here. However, this method _will_ call out to EPR on every single invocation in order to resolve the `latest` version of the given package: https://github.com/elastic/kibana/blob/5db8bcb4b051d1bd695cce7f8e1cd610a6894bf3/x-pack/plugins/fleet/server/services/epm/packages/get.ts#L139-L142

This is a _lot_ of overhead to incur every time we need a `PackageInfo` object - which happens to be just about all the time. 

One thing that would make this a little bit easier would be a more complete in-memory caching solution that would prevent these network calls, or allow us to simply pull the `PackageInfo` objects directly from a key/value store. However that would be a pretty large undertaking, I think, so I opted in this PR to simply add some memoization logic to a few key workflows where we request a `PackageInfo` value in a loop.